btcec/nfield_test.go

```diff
@@ -0,0 +1,925 @@
+	_, err := rand.Read(data)
+	if err != nil {
+		t.Fatalf("failed to read random data")
+	}
+	N := btcec.S256().N
+	a := new(big.Int).SetBytes(data)
+	aN := btcec.NewNfieldVal().SetByteSlice(a.Bytes())
+	for i := 0; i < 100; i++ {
+		c := big.NewInt(0)
+		c.Mul(a, a)
+		c.Mod(c, N)
+		cN := btcec.NewNfieldVal()
+		cN.SquareVal(aN)
+		cstr := fmt.Sprintf("%064x", c)
+		if cN.String() != cstr {
+			t.Fatalf("expected strings to be the same, was not: %v %v", cN.String(), cstr)
+		}
+		a = c
+		aN = cN
+	}
+}
+
+// TestInverse ensures that finding the multiplicative inverse via Inverse works
+// as expected.
+func TestNfieldInverse(t *testing.T) {
+	tests := []struct {
+		in       string // hex encoded value
+		expected string // expected hex encoded value
+	}{
+		// secp256k1 n (aka 0)
+		{"0", "0"},
+		{"fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141", "0"},
+		{"0", "fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141"},
+		// secp256k1 n-1
+		{
+			"fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364140",
+			"fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364140",
+		},
+		// secp256k1 n-2
+		{
+			"fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd036413f",
+			"7fffffffffffffffffffffffffffffff5d576e7357a4501ddfe92f46681b20a0",
+		},
+		// Random sampling
+		{
+			"16fb970147a9acc73654d4be233cc48b875ce20a2122d24f073d29bd28805aca",
+			"bfd13908390c647c650937b08b923060e0ef10534f0b19d5904ea3d3ffd20729",
+		},
+		{
+			"69d1323ce9f1f7b3bd3c7320b0d6311408e30281e273e39a0d8c7ee1c8257919",
+			"8d5e9ed9c17b1801268f83b1d1ab3422672d37c060885e47dba523dc20a00774",
+		},
+		{
+			"e0debf988ae098ecda07d0b57713e97c6d213db19753e8c95aa12a2fc1cc5272",
+			"1b1509cf12e6e94c6f441d61c25adc9f7110008e1a192f95ca1ca0f3b60e4e06",
+		},
+		{
+			"dcd394f91f74c2ba16aad74a22bb0ed47fe857774b8f2d6c09e28bfb14642878",
+			"dc8d7bebb5e528bacaf73e4d97830df032b813e7f63359c2baa2ddc355c2f53c",
+		},
+	}
+
+	t.Logf("Running %d tests", len(tests))
+	for i, test := range tests {
+		f := btcec.NewNfieldVal().SetHex(test.in).Normalize()
+		expected := btcec.NewNfieldVal().SetHex(test.expected).Normalize()
+		result := f.Inverse().Normalize()
+		if !result.Equals(expected) {
+			t.Errorf("nfieldVal.Inverse #%d wrong result\n"+
+				"got: %v\nwant: %v", i, result, expected)
+			continue
+		}
+	}
+}
+
+// TestNfieldInvRand tests the inverse correctness by doing the
+// same calculation with Big.Int
+func TestNfieldInverseRand(t *testing.T) {
+	N := btcec.S256().N
+	for i := 0; i < 100; i++ {
+		data := make([]byte, 32)
+		_, err := rand.Read(data)
+		if err != nil {
+			t.Fatalf("failed to read random data")
+		}
+		a := new(big.Int).SetBytes(data)
+		aN := btcec.NewNfieldVal().SetByteSlice(a.Bytes())
+		c := big.NewInt(0)
+		c.ModInverse(a, N)
+		cN := aN.Inverse()
+		cstr := fmt.Sprintf("%064x", c)
+		if cN.String() != cstr {
+			t.Fatalf("expected strings to be the same, was not: %v %v", cN.String(), cstr)
```

Comment:
I'll avoid repeating this comment for the rest, but there are several more cases that this applies to.